### PR TITLE
SOLR-17194: CloudHttp2SolrClient doesn't properly use httpClient Builder on certain paths

### PR DESCRIPTION
**Pull request by epugh**
https://issues.apache.org/jira/browse/SOLR-17194

# Description

Discovered a bug in the use of Export Tool with Basic Auth parameters doesn't create a proper CloudHttp2SolrClient.  

Undercovered that we pass in a httpClient builder to CloudHttp2SolrClient, but in at least this code path we never use, so we have a null httpClient, which means we build a default one, which means if you passed one in then you don't get your Basic Auth params!

# Solution

A test and a small fix.   I did not try to reasoon about all the builder stuff..  I find the combination of httpClient, and the solrClient and their various builders VERY hard to understand...   I wrote at test, I fixed the code, the test passes...

# Tests

Added a test to the Export Tool

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by epugh** [comment]
Pinged community for feedback, I'd like to merge this week...

File: solr/core/src/test/org/apache/solr/cli/TestExportTool.java [hunk 1]
```diff
@@ -218,6 +231,66 @@ public void testVeryLargeCluster() throws Exception {
...
+                    singletonMap(USER, getSaltedHashedValue(PASS)))));
+
+    configureCluster(2)
+        .addConfig("conf", configset("cloud-minimal"))
+        .withSecurityJson(SECURITY_JSON)
+        .configure();
+
+    CloudSolrClient cloudSolrClient = cluster.getSolrClient();
+
+    CollectionAdminRequest.createCollection(COLLECTION_NAME, "conf", 8, 1)
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/cli/TestExportTool.java:266]: 8 shards?!
To tame our tests, tests should have a comment justifying >1 nodes or >1 shards.  If there is a reason; justify it.

Comment (epugh) [reply solr/core/src/test/org/apache/solr/cli/TestExportTool.java:266]: total copy paste ;-)

Comment (epugh) [reply solr/core/src/test/org/apache/solr/cli/TestExportTool.java:266]: Oh, wait...  I found it..  The test is `testVeryLargeCluster` and it's a nightly...   so I think the is why....

Comment (epugh) [reply solr/core/src/test/org/apache/solr/cli/TestExportTool.java:266]: do you want another specific comment on the test....

**Comment by epugh** [comment]
@dsmiley do the rest of the changes look okay to you?

File: solr/core/src/test/org/apache/solr/cli/TestExportTool.java (hunk 1 above)
Comment (epugh) [reply solr/core/src/test/org/apache/solr/cli/TestExportTool.java]: I made the change to reduce the number...

File: solr/core/src/test/org/apache/solr/cli/TestExportTool.java [hunk 2]
```diff
@@ -218,6 +231,66 @@ public void testVeryLargeCluster() throws Exception {
...
+                "authentication",
+                Map.of(
+                    "class",
+                    BasicAuthPlugin.class.getName(),
+                    "blockUnknown",
+                    true,
+                    "credentials",
+                    singletonMap(USER, getSaltedHashedValue(PASS)))));
+
+    configureCluster(2)
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/cli/TestExportTool.java:259]: Can the behavior be tested with one node instead of two?  Authentication seems unrelated to the size of the cluster.

File: solr/core/src/test/org/apache/solr/cli/TestExportTool.java [hunk 3]
```diff
@@ -218,6 +231,66 @@ public void testVeryLargeCluster() throws Exception {
...
+        .addConfig("conf", configset("cloud-minimal"))
+        .withSecurityJson(SECURITY_JSON)
+        .configure();
+
+    CloudSolrClient cloudSolrClient = cluster.getSolrClient();
+
+    CollectionAdminRequest.createCollection(COLLECTION_NAME, "conf", 2, 1)
+        .setBasicAuthCredentials(USER, PASS)
+        .process(cluster.getSolrClient());
+    cluster.waitForActiveCollection(COLLECTION_NAME, 8, 8);
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/cli/TestExportTool.java:269]: You won't get 8 replicas anymore; you changed this.  The test probably fails now.  I hate `waitForActiveCollection`; the collection is already active when createCollection finishes.  The risk this method mitigates is that the SolrClient returned by `cluster.getSolrClient` doesn't know it yet, but this test doesn't use the client henceforth (unlike most tests).

File: solr/core/src/test/org/apache/solr/cli/TestExportTool.java [hunk 4]
```diff
@@ -218,6 +231,66 @@ public void testVeryLargeCluster() throws Exception {
...
+                    singletonMap(USER, getSaltedHashedValue(PASS)))));
+
+    configureCluster(2)
+        .addConfig("conf", configset("cloud-minimal"))
+        .withSecurityJson(SECURITY_JSON)
+        .configure();
+
+    CloudSolrClient cloudSolrClient = cluster.getSolrClient();
+
+    CollectionAdminRequest.createCollection(COLLECTION_NAME, "conf", 2, 1)
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/cli/TestExportTool.java:266]: Two shards doesn't seem pertinent to what is being tested.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java [hunk 5]
```diff
@@ -422,7 +422,10 @@ public CloudHttp2SolrClient build() {
           }
         } else if (!solrUrls.isEmpty()) {
           try {
-            stateProvider = new Http2ClusterStateProvider(solrUrls, httpClient);
+            if (this.internalClientBuilder != null) {
+              this.httpClient = internalClientBuilder.build();
```
Comment (stillalex) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:426]: I think there might be an issue here with the `httpClient` instance created here is not being closed correctly. not 100% sure I will need to check more closely. before this change I agree there was a default client being created lacking the correct params but it also closed.

Comment (epugh) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:426]: thanks for chiming in...  so, this change is making my TestExportTool blow up with unclosed httpclients..   I've banged my head on this for a LONG time, and not making headway.   What I can say is that the test with the BasicAuth demonstrates the problem, but I can't grok the soluiton...   @stillalex I would love your ideas if you wanted to take a stab at it..   it felt like such an easy fix!

Comment (laminelam) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:426]: That's funny, I pushed 2 weeks ago an internal PR (at my work) for this exact same bug, as I faced it myself.

@stillalex  is correct, the resource won't be released if an exception is raised when instantiating the  _Http2ClusterStateProvider_

Plus we end up creating two _httpCient_ instances. CloudHttp2SolrClient and stateProvider would use 2 different _httClient_s.

In my PR, I pushed most of the code into the CloudHttp2SolrClient constructor. I added a test case to test if the same _httpClient_ is preserved in both _CloudHttp2SolrClient_ and _Http2ClusterStateProvider_.
All test cases passed, including _TestExportTool_.

Will see if I can push the PR here (upstream).

Comment (epugh) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:426]: Thanks all for looking at this.  I have definitly hit the "end" of my rope on this one, just can't figure it out...   My Java-fu isn't up to this one.   

I am inclined to close this PR in favour of what you can whip up..  However, since I really really really really really really want basic auth to work on all the tools, I am hesitating to actually do that!

Comment (laminelam) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:426]: Actually I have 2 PRs that I am pretty sure would help a lot on this.
One that fixes this bug.
And anther one that I talked about [here](https://github.com/apache/solr/pull/1994#discussion_r1535512663) where I refactored ZK related SolrCli tools to separate _SolrZkClient_ and _CloudHttp2SolrClient_  instantiation/usage.

Will see if I can speed up the internal review and push them upstream.

Comment (laminelam) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java:426]: @epugh 
Here's the [PR](https://github.com/apache/solr/pull/2412)
Tested (locally) your testWithBasicAuth() code and the test passed

**Comment by epugh** [comment]
Closing in favour of #2412
